### DISCUSSION TITLE
Proposal: Narrowly-scoped pattern variables with let clause

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#5632] [Language Ideas]
The decision to make the scope of variable patterns leak has always been a controversial one.  While it does enable some use cases without requiring further language features it does so at the expense of unexpected scoping rules and identifier collisions.  The latter has been problematic enough for the LDM to have thrown around some ideas as to how the language might allow for the same variable pattern identifier to exist within overlapping scopes but only if they are not definitely assigned.

I'd like to explore the idea of perhaps using the `let` clause as a new form of variable pattern (or as a modifier to the existing pattern) which would change the scoping rules of the introduced identifier:

```cs
if (person is Student { Name: let name })
{
    Console.WriteLine($"Welcome to class, {name.FirstName}!"); 
}
else if (person is Professor { Name: let name })
{
    Console.WriteLine($"Good morning Professor {name.LastName}!");
}
// name not in scope here
```

or, with simple types:

```cs
if (obj is let int value)
{
    // use value here
}
else if (obj is let string value)
{
    // use value here
}
// value not in scope here
```

### Proposed Grammar
```
pattern
    : declaration_pattern
    | var_pattern
    | range_declaration_pattern
    | // all other pattern forms
    ;

declaration_pattern
    : type simple_designation
    ;

var_pattern
    : 'var' simple_designation
    ;

range_declaration_pattern
    : typed_range_declaration_pattern
    | let_pattern
    ;

typed_range_declaration_pattern
    : 'let' type simple_designation
    ;

let_pattern
    : 'let' simple_designation
    ;
```

Questions:
1. Would the `let` pattern be a separate pattern, a counterpoint to `var`, or would it be a modifier to the existing patterns?
2. Is it worth debating if the variable should be implicitly `readonly`, as range variables are and as suggested for the `let` shorthand for [`readonly` locals](https://github.com/dotnet/csharplang/issues/188)?

**Comment by CyrusNajmabadi** [comment]
This is one approach that could work.  However, i'm also amenable to just relaxing the existing rules to allow for this:

```c#
if (person is Student { Name: var name })
{
    Console.WriteLine($"Welcome to class, {name.FirstName}!"); 
}
else if (person is Professor { Name: var name })
{
    Console.WriteLine($"Good morning Professor {name.LastName}!");
}
```

This could be done with varying levels of complexity.  For example, `var name` is not definitely assigned in the 'else' block, so we could allow a new variable to be introduced there.  

This would effectively be allowing an error case to become non-error in some cases. 

Where it gets tricker (but what i would still like to support) are cases like:

```c#
... switch { sbyte v => ..., int v => ..., long v => ... }
```

Here not only is the name not definitely assigned in the other cases, it also has a different type entirely.  I would also like this to work.

**Reply by HaloFour** [reply]
> Where it gets tricker (but what i would still like to support) are cases like

Right, I am keeping both of the use cases in mind with this proposal.  Rather than allowing for the variable declarations to overlap by virtue of definite assignment, which is useful for disjunctive patterns, but not if the type of the variable might be different.  This proposal would instead involve completely separate and unrelated declarations that just happen to have the same identifier and scope.

I imagine that it would be possible for the compiler to work that out, but I think it would still be pretty confusing:

```cs
if (obj is int value)
{
    // use value here
}
else if (obj is string value)
{
    // use value here
}

// what is value here?
{
    // can I redeclare value here?
}
```

**Reply by RikkiGibson** [reply]
I feel like with Cyrus's idea, the problem is just figuring out how to adjust the local variable shadowing rules to make it legal.

**Reply by 333fred** [reply]
Note that this separate proposal is effectively https://github.com/dotnet/csharplang/issues/4018. I also think it's orthogonal to narrowly-scoped variables. Narrowly-scoped variables would be trivial reusable because they are not in scope in outside, whereas this is really about allowing redeclaration.

**Reply by HaloFour** [reply]
@RikkiGibson 

I don't disagree and I think it's worth exploring.  I think it will be a bit strange if the compiler implicitly could handle both the leaky disjunctive case as well as the overlapping/shadowing identifier case.  I don't think the questions in the code example above could have a single simple answer.  But maybe that's not really an issue in practice.

That said, there may be further value in having a way to explicitly opt-in to a narrow scope (and possibility readonly-ness), both as it solves those cases as well as it scratches the itch of those developers who have never been happy with the leaky scope rules.

**Reply by RikkiGibson** [reply]
For the disjunctive case it could be worth looking at giving patterns the ability to "write out" to existing variables, be they pattern variables or not. I don't recall if this was already considered and rejected.

strawman:
```cs
if (MyGetStringMethod1() is string myString
    or MyGetStringMethod2() is out myString)
{
}
```

**Reply by orthoxerox** [reply]
@CyrusNajmabadi You suggestion covers disjoint scopes *and* scopes where you can Φ two previous scopes together, which is sometimes exactly what I need and sometimes exactly what I don't need. ¿Por qué no los dos?

1. Disjoint scopes can have the same mutable variable declared in both as long as its type is the same.
2. let-bindings have a narrow scope. Or they could just be shadowable, Rust-style. I've had a few cases where I just couldn't come up with a good name for a (not even necessarily pattern) variable, especially for unwrapped wrapper types. Being able to just discard the previous let-binding with the same name would be pretty convenient.

**Reply by alrz** [reply]
> ```cs
> if (obj is int value) ;
> else if (obj is string value) ;
> ```

https://github.com/dotnet/csharplang/pull/5781 covers this except that types for `value` must match. However, in this example the two are completely independent. I think we should be able to somehow relax the rules to allow this. cc @333fred

**Reply by HaloFour** [reply]
@alrz 

Theoretically I love that idea, but it doesn't seem feasible unless such a declaration also narrows the scope of the identifier:

```cs
if (obj is int value) { ... }
else if (obj is string value) { ... }

// value is in scope here, but what type is it?
```

**Reply by alrz** [reply]
@HaloFour 

You're right in that this is no longer "redeclaration" and should be considered "shadowing" which complicates things a bit since we have two independent locals now. I guess the correct error for `value` usages after `if` would be "ambiguous reference" so neither read nor reassignment would be valid.

**Reply by 333fred** [reply]
Agreed with Halo. I'm not supportive of it.

**Comment by iam3yal** [comment]
Would it support the case like in the example below? like shadowing `prime`.
```csharp
public static IEnumerable<int> PrimeFactor(int input)
{
    // ...

    return PrimeFactorImpl(input);

    static IEnumerable<int> PrimeFactorImpl(int input, int prime = 2)
    {
        // ...

        foreach (var returnedPrime in PrimeFactorImpl(input, prime))
        {
            yield return returnedPrime;
        }
    }
}
 ```

**Reply by HaloFour** [reply]
I don't see pattern matching used in that example?

As for whether a clause like this would behave any differently than normal shadowing rules I think that's up for discussion.

**Reply by iam3yal** [reply]
Yeah there isn't pattern matching here but I thought to bring this up because at times it's really annoying to come up with variables names so I thought that `let` might be helpful in this case as well where you could do `let prime in PrimeFactorImpl(input, prime)` but I'm not sure.
